diff -Nur zstd-1.5.4/examples/Makefile zstd/examples/Makefile
--- zstd-1.5.4/examples/Makefile	2023-02-10 08:41:50.000000000 +0800
+++ zstd/examples/Makefile	2026-01-06 19:39:32.187002580 +0800
@@ -12,6 +12,8 @@
 CPPFLAGS += -I$(LIBDIR)
 LIB = $(LIBDIR)/libzstd.a
 
+KAEZSTD_LDFLAGS=-L/usr/local/lib -L/usr/local/kaezstd/lib -Wl,-rpath,/usr/local/lib:/usr/local/kaezstd/lib -lzstd -lwd_comp -lkaezstd
+LDFLAGS+=$(KAEZSTD_LDFLAGS)
 
 .PHONY: default
 default: all
diff -Nur zstd-1.5.4/lib/common/error_private.c zstd/lib/common/error_private.c
--- zstd-1.5.4/lib/common/error_private.c	2023-02-10 08:41:50.000000000 +0800
+++ zstd/lib/common/error_private.c	2026-01-06 19:39:32.187002580 +0800
@@ -12,7 +12,7 @@
 
 #include "error_private.h"
 
-const char* ERR_getErrorString(ERR_enum code)
+__attribute__ ((visibility("default"))) const char* ERR_getErrorString(ERR_enum code)
 {
 #ifdef ZSTD_STRIP_ERROR_STRINGS
     (void)code;
diff -Nur zstd-1.5.4/lib/common/error_private.h zstd/lib/common/error_private.h
--- zstd-1.5.4/lib/common/error_private.h	2023-02-10 08:41:50.000000000 +0800
+++ zstd/lib/common/error_private.h	2026-01-06 19:39:32.187002580 +0800
@@ -68,7 +68,7 @@
 *  Error Strings
 ******************************************/
 
-const char* ERR_getErrorString(ERR_enum code);   /* error_private.c */
+__attribute__ ((visibility("default"))) const char* ERR_getErrorString(ERR_enum code);   /* error_private.c */
 
 ERR_STATIC const char* ERR_getErrorName(size_t code)
 {
diff -Nur zstd-1.5.4/lib/compress/zstd_compress.c zstd/lib/compress/zstd_compress.c
--- zstd-1.5.4/lib/compress/zstd_compress.c	2023-02-10 08:41:50.000000000 +0800
+++ zstd/lib/compress/zstd_compress.c	2026-01-07 16:14:28.099002580 +0800
@@ -28,6 +28,10 @@
 #include "zstd_compress_superblock.h"
 #include  "../common/bits.h"      /* ZSTD_highbit32 */
 
+#ifdef CONF_KAEZSTD
+#include "kaezstd.h"
+#endif
+
 /* ***************************************************************
 *  Tuning parameters
 *****************************************************************/
@@ -107,6 +111,15 @@
         assert(!ZSTD_isError(err));
         (void)err;
     }
+    #ifdef CONF_KAEZSTD
+        int ret ;
+        ret = kaezstd_init(cctx);
+        if (ret == 0) {
+            cctx->kaeInited = 1;
+        } else {
+            cctx->kaeInited = 0;
+        }
+    #endif
 }
 
 ZSTD_CCtx* ZSTD_createCCtx_advanced(ZSTD_customMem customMem)
@@ -173,6 +186,11 @@
     ZSTDMT_freeCCtx(cctx->mtctx); cctx->mtctx = NULL;
 #endif
     ZSTD_cwksp_free(&cctx->workspace, cctx->customMem);
+#ifdef CONF_KAEZSTD
+    if (cctx->kaeInited == 1) {
+        kaezstd_release(cctx);
+    }
+#endif
 }
 
 size_t ZSTD_freeCCtx(ZSTD_CCtx* cctx)
@@ -721,6 +739,9 @@
         break;
 
     case ZSTD_c_compressionLevel:
+#ifdef CONF_KAEZSTD
+	cctx->kaeLevel = value;
+#endif
     case ZSTD_c_windowLog:
     case ZSTD_c_hashLog:
     case ZSTD_c_chainLog:
@@ -1325,6 +1346,9 @@
  *  Also dumps dictionary */
 size_t ZSTD_CCtx_reset(ZSTD_CCtx* cctx, ZSTD_ResetDirective reset)
 {
+#ifdef CONF_KAEZSTD
+    if (cctx->kaeInited == 1) { kaezstd_reset(cctx); }
+#endif
     if ( (reset == ZSTD_reset_session_only)
       || (reset == ZSTD_reset_session_and_parameters) ) {
         cctx->streamStage = zcss_init;
@@ -2076,6 +2100,9 @@
                 RETURN_ERROR_IF(zc->blockState.prevCBlock == NULL, memory_allocation, "couldn't allocate prevCBlock");
                 zc->blockState.nextCBlock = (ZSTD_compressedBlockState_t*) ZSTD_cwksp_reserve_object(ws, sizeof(ZSTD_compressedBlockState_t));
                 RETURN_ERROR_IF(zc->blockState.nextCBlock == NULL, memory_allocation, "couldn't allocate nextCBlock");
+            #ifdef CONF_KAEZSTD
+        		ZSTD_memset(zc->blockState.nextCBlock, 0, ZSTD_cwksp_align(sizeof(ZSTD_compressedBlockState_t), sizeof(void*)));
+            #endif
                 zc->entropyWorkspace = (U32*) ZSTD_cwksp_reserve_object(ws, ENTROPY_WORKSPACE_SIZE);
                 RETURN_ERROR_IF(zc->entropyWorkspace == NULL, memory_allocation, "couldn't allocate entropyWorkspace");
         }   }
@@ -3126,6 +3153,15 @@
             ms->nextToUpdate = curr - MIN(192, (U32)(curr - ms->nextToUpdate - 384));
     }
 
+#ifdef CONF_KAEZSTD
+    if (zc->kaeInited == 1) {
+        int ret;
+        ret = kaezstd_compress(zc, src, srcSize);
+        if (ret) {
+            return ZSTDbss_noCompress;
+        }
+    } else {
+#endif
     /* select and store sequences */
     {   ZSTD_dictMode_e const dictMode = ZSTD_matchState_dictMode(ms);
         size_t lastLLSize;
@@ -3248,6 +3284,9 @@
         {   const BYTE* const lastLiterals = (const BYTE*)src + srcSize - lastLLSize;
             ZSTD_storeLastLiterals(&zc->seqStore, lastLiterals, lastLLSize);
     }   }
+#ifdef CONF_KAEZSTD
+    }
+#endif
     return ZSTDbss_compress;
 }
 
@@ -4370,7 +4409,11 @@
 
         /* Ensure hash/chain table insertion resumes no sooner than lowlimit */
         if (ms->nextToUpdate < ms->window.lowLimit) ms->nextToUpdate = ms->window.lowLimit;
-
+#ifdef CONF_KAEZSTD
+        if (cctx->kaeInited == 1) {
+            cctx->kaeFrameMode = lastBlock;
+        }
+#endif
         {   size_t cSize;
             if (ZSTD_useTargetCBlockSize(&cctx->appliedParams)) {
                 cSize = ZSTD_compressBlock_targetCBlockSize(cctx, op, dstCapacity, ip, blockSize, lastBlock);
@@ -4389,6 +4432,11 @@
 
                 if (cSize == 0) {  /* block is not compressible */
                     cSize = ZSTD_noCompressBlock(op, dstCapacity, ip, blockSize, lastBlock);
+                #ifdef CONF_KAEZSTD
+                    if (cctx->kaeInited == 1) {
+                        kaezstd_setstatus(cctx, TUPLE_STATUS_NOCOMPRESS);
+                    }
+                #endif
                     FORWARD_IF_ERROR(cSize, "ZSTD_noCompressBlock failed");
                 } else {
                     U32 const cBlockHeader = cSize == 1 ?
@@ -4396,6 +4444,11 @@
                         lastBlock + (((U32)bt_compressed)<<1) + (U32)(cSize << 3);
                     MEM_writeLE24(op, cBlockHeader);
                     cSize += ZSTD_blockHeaderSize;
+                #ifdef CONF_KAEZSTD
+                    if (cctx->kaeInited == 1) {
+                        kaezstd_setstatus(cctx, cSize == 1 ? TUPLE_STATUS_RLEBLOCK : TUPLE_STATUS_COMPRESS);
+                    }
+                #endif
                 }
             }  /* if (ZSTD_useTargetCBlockSize(&cctx->appliedParams))*/
 
@@ -4997,6 +5050,9 @@
     ZSTD_CCtx_params cctxParams;
     {   ZSTD_parameters const params = ZSTD_getParams_internal(compressionLevel, ZSTD_CONTENTSIZE_UNKNOWN, dictSize, ZSTD_cpm_noAttachDict);
         ZSTD_CCtxParams_init_internal(&cctxParams, &params, (compressionLevel == 0) ? ZSTD_CLEVEL_DEFAULT : compressionLevel);
+    #ifdef CONF_KAEZSTD
+    	cctx->kaeLevel = compressionLevel;
+    #endif
     }
     DEBUGLOG(4, "ZSTD_compressBegin_usingDict (dictSize=%u)", (unsigned)dictSize);
     return ZSTD_compressBegin_internal(cctx, dict, dictSize, ZSTD_dct_auto, ZSTD_dtlm_fast, NULL,
@@ -5098,6 +5154,9 @@
             (unsigned)cctx->consumedSrcSize);
     }
     ZSTD_CCtx_trace(cctx, endResult);
+#ifdef CONF_KAEZSTD
+    kaezstd_reset(cctx);
+#endif
     return cSize + endResult;
 }
 
@@ -5142,6 +5201,9 @@
         ZSTD_parameters const params = ZSTD_getParams_internal(compressionLevel, srcSize, dict ? dictSize : 0, ZSTD_cpm_noAttachDict);
         assert(params.fParams.contentSizeFlag == 1);
         ZSTD_CCtxParams_init_internal(&cctx->simpleApiParams, &params, (compressionLevel == 0) ? ZSTD_CLEVEL_DEFAULT: compressionLevel);
+    #ifdef CONF_KAEZSTD
+    	cctx->kaeLevel = compressionLevel;
+    #endif
     }
     DEBUGLOG(4, "ZSTD_compress_usingDict (srcSize=%u)", (unsigned)srcSize);
     return ZSTD_compress_advanced_internal(cctx, dst, dstCapacity, src, srcSize, dict, dictSize, &cctx->simpleApiParams);
diff -Nur zstd-1.5.4/lib/compress/zstd_compress_internal.h zstd/lib/compress/zstd_compress_internal.h
--- zstd-1.5.4/lib/compress/zstd_compress_internal.h	2023-02-10 08:41:50.000000000 +0800
+++ zstd/lib/compress/zstd_compress_internal.h	2026-01-06 19:39:32.191002580 +0800
@@ -18,6 +18,10 @@
 /*-*************************************
 *  Dependencies
 ***************************************/
+#ifdef CONF_KAEZSTD
+#include <stdint.h>
+#endif
+
 #include "../common/zstd_internal.h"
 #include "zstd_cwksp.h"
 #ifdef ZSTD_MULTITHREAD
@@ -401,6 +405,12 @@
 } ZSTD_externalMatchCtx;
 
 struct ZSTD_CCtx_s {
+#ifdef CONF_KAEZSTD
+    unsigned char kaeInited;
+    unsigned int kaeFrameMode;
+    uintptr_t kaeConfig;
+    int kaeLevel;
+#endif
     ZSTD_compressionStage_e stage;
     int cParamsChanged;                  /* == 1 if cParams(except wlog) or compression level are changed in requestedParams. Triggers transmission of new params to ZSTDMT (if available) then reset to 0. */
     int bmi2;                            /* == 1 if the CPU supports BMI2 and 0 otherwise. CPU support is determined dynamically once per context lifetime. */
diff -Nur zstd-1.5.4/lib/libzstd.mk zstd/lib/libzstd.mk
--- zstd-1.5.4/lib/libzstd.mk	2023-02-10 08:41:50.000000000 +0800
+++ zstd/lib/libzstd.mk	2026-01-06 19:39:32.191002580 +0800
@@ -107,6 +107,11 @@
 LDFLAGS  += $(MOREFLAGS)
 FLAGS     = $(CPPFLAGS) $(CFLAGS) $(ASFLAGS) $(LDFLAGS)
 
+KAEZSTD_CFLAGS=-I/usr/local/kaezstd/include -I$(KAEZSTDPATH)/include -I. -I../lib -DCONF_KAEZSTD
+KAEZSTD_LDFLAGS=-L/usr/local/kaezstd/lib -L$(KAEBUILDPATH)/lib -L$(KAEZSTDPATH) -Wl,-rpath,/usr/local/kaezstd/lib:/usr/local/lib:$(KAEBUILDPATH)/lib -l:libkaezstd.so.2.2.0
+CFLAGS+=$(KAEZSTD_CFLAGS)
+LDFLAGS+=$(KAEZSTD_LDFLAGS)
+
 ifndef ALREADY_APPENDED_NOEXECSTACK
 export ALREADY_APPENDED_NOEXECSTACK := 1
 ifeq ($(shell echo "int main(int argc, char* argv[]) { (void)argc; (void)argv; return 0; }" | $(CC) $(FLAGS) -z noexecstack -x c -Werror - -o $(VOID) 2>$(VOID) && echo 1 || echo 0),1)
diff -Nur zstd-1.5.4/programs/Makefile zstd/programs/Makefile
--- zstd-1.5.4/programs/Makefile	2023-02-10 08:41:50.000000000 +0800
+++ zstd/programs/Makefile	2026-06-22 19:00:00.000000000 +0800
@@ -137,6 +137,7 @@
 zstd : CPPFLAGS += $(THREAD_CPP) $(ZLIBCPP) $(LZMACPP) $(LZ4CPP)
 zstd : LDFLAGS += $(THREAD_LD) $(DEBUGFLAGS_LD)
 zstd : LDLIBS += $(ZLIBLD) $(LZMALD) $(LZ4LD)
+zstd : LDLIBS += $(KAEZSTD_LDFLAGS)
 zstd : CPPFLAGS += -DZSTD_LEGACY_SUPPORT=$(ZSTD_LEGACY_SUPPORT)
 ifneq (,$(filter Windows%,$(OS)))
 zstd : $(RES_FILE)
