diff --git a/uadk/v1/drv/hisi_sec_udrv.c b/uadk/v1/drv/hisi_sec_udrv.c
index d046327..0b3e4f5 100644
--- a/uadk/v1/drv/hisi_sec_udrv.c
+++ b/uadk/v1/drv/hisi_sec_udrv.c
@@ -186,6 +186,9 @@ static int fill_cipher_bd2_mode(struct wcrypto_cipher_msg *msg,
 	case WCRYPTO_CIPHER_XTS:
 		sqe->type2.c_mode = C_MODE_XTS;
 		break;
+	case WCRYPTO_CIPHER_CTR:
+		sqe->type2.c_mode = C_MODE_CTR;
+		break;
 	case WCRYPTO_CIPHER_CBC_CS1:
 		sqe->type2.c_mode = C_MODE_CBC_CS;
 		sqe->type2.c_width = C_WIDTH_CS1;
