From e4eb46f7bd5b1a1c576d548dd2265621b4a6ac9a Mon Sep 17 00:00:00 2001
From: LiuYongYang <liuyongyang@huawei.com>
Date: Thu, 27 Jun 2024 15:05:43 +0800
Subject: [PATCH] fix uadk zstd bug

---
 uadk/v1/drv/hisi_zip_udrv.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/uadk/v1/drv/hisi_zip_udrv.c b/uadk/v1/drv/hisi_zip_udrv.c
index dfd1fb3..f161c32 100644
--- a/uadk/v1/drv/hisi_zip_udrv.c
+++ b/uadk/v1/drv/hisi_zip_udrv.c
@@ -501,8 +501,8 @@ static int fill_zip_addr_lz77_zstd(void *ssqe,
 	} else {
 		sqe->cipher_key_addr_l = lower_32_bits((__u64)addr.dest_addr);
 		sqe->cipher_key_addr_h = upper_32_bits((__u64)addr.dest_addr);
-		sqe->dest_addr_l = lower_32_bits((__u64)addr.dest_addr + msg->in_size);
-		sqe->dest_addr_h = upper_32_bits((__u64)addr.dest_addr + msg->in_size);
+		sqe->dest_addr_l = lower_32_bits((__u64)addr.dest_addr + msg->in_size + ZSTD_LIT_RSV_SIZE);
+		sqe->dest_addr_h = upper_32_bits((__u64)addr.dest_addr + msg->in_size + ZSTD_LIT_RSV_SIZE);
 	}
 
 	sqe->stream_ctx_addr_l = lower_32_bits((__u64)addr.ctxbuf_addr);
@@ -672,7 +672,7 @@ static void fill_priv_lz77_zstd(void *ssqe, struct wcrypto_comp_msg *recv_msg)
 		format->sequences_start = zstd_out->sequence;
 	} else {
 		format->literals_start = recv_msg->dst;
-		format->sequences_start = recv_msg->dst + recv_msg->in_size;
+		format->sequences_start = recv_msg->dst + recv_msg->in_size + ZSTD_LIT_RSV_SIZE;
 		format->freq = (void *)(&format->lit_length_overflow_pos + 1);
 	}
 
-- 
2.33.0

