diff -Naru a/uadk/cleanup.sh b/uadk/cleanup.sh
--- a/uadk/cleanup.sh	2024-12-20 11:13:50.000000000 +0800
+++ b/uadk/cleanup.sh	2024-12-20 11:15:59.679983640 +0800
@@ -10,9 +10,7 @@
        ar-lib m4 Makefile.in missing src/Makefile src/Makefile.in	     \
        test/Makefile test/Makefile.in test/hisi_hpre_test/Makefile.in	     \
        test/hisi_hpre_test/Makefile test/hisi_sec_test/Makefile              \
-       test/hisi_sec_test/Makefile.in test/hisi_zip_test/Makefile            \
-       test/hisi_zip_test/Makefile.in uadk_tool/Makefile            \
-       uadk_tool/Makefile.in         \
+       test/hisi_sec_test/Makefile.in            \
        v1/Makefile.in v1/Makefile.in v1/test/Makefile v1/test/Makefile.in    \
        v1/test/test_mm/Makefile v1/test/test_mm/Makefile.in		     \
        v1/test/bmm_test/Makefile v1/test/bmm_test/Makefile.in"
diff -Naru a/uadk/configure.ac b/uadk/configure.ac
--- a/uadk/configure.ac	2024-12-20 11:13:50.000000000 +0800
+++ b/uadk/configure.ac	2024-12-20 11:15:59.679983640 +0800
@@ -91,8 +91,6 @@
 		 lib/libwd_crypto.pc lib/libwd_comp.pc lib/libwd.pc
 		 test/Makefile
 		 test/hisi_hpre_test/Makefile
-		 test/hisi_zip_test/Makefile
-		 uadk_tool/Makefile
 		 sample/Makefile
 		 v1/test/Makefile
 		 v1/test/bmm_test/Makefile
@@ -101,6 +99,5 @@
 		 v1/test/hisi_trng_test/Makefile
 		 v1/test/hisi_sec_test/Makefile
 		 v1/test/hisi_sec_test_sgl/Makefile
-		 v1/test/hisi_zip_test/Makefile
 		 v1/test/hisi_zip_test_sgl/Makefile])
 AC_OUTPUT
diff -Naru a/uadk/Makefile.am b/uadk/Makefile.am
--- a/uadk/Makefile.am	2024-12-20 11:13:50.000000000 +0800
+++ b/uadk/Makefile.am	2024-12-20 11:15:59.679983640 +0800
@@ -138,4 +138,3 @@
 pkgconfig_DATA = lib/libwd_crypto.pc lib/libwd_comp.pc lib/libwd.pc
 CLEANFILES += $(pkgconfig_DATA)
 
-SUBDIRS=. test v1/test uadk_tool sample
diff -Naru a/uadk/test/Makefile.am b/uadk/test/Makefile.am
--- a/uadk/test/Makefile.am	2024-12-20 11:13:50.000000000 +0800
+++ b/uadk/test/Makefile.am	2024-12-20 11:15:59.679983640 +0800
@@ -16,8 +16,5 @@
 if HAVE_CRYPTO
 SUBDIRS += hisi_hpre_test
 
-if HAVE_ZLIB
-SUBDIRS += hisi_zip_test
-endif
 
 endif
diff -Naru a/uadk/v1/test/Makefile.am b/uadk/v1/test/Makefile.am
--- a/uadk/v1/test/Makefile.am	2024-12-20 11:13:50.000000000 +0800
+++ b/uadk/v1/test/Makefile.am	2024-12-20 11:15:59.679983640 +0800
@@ -2,9 +2,6 @@
 
 SUBDIRS=. test_mm bmm_test
 
-if HAVE_ZLIB
-SUBDIRS+=hisi_zip_test
-endif
 
 SUBDIRS+=hisi_zip_test_sgl
 SUBDIRS+=hisi_trng_test
