diff -urN a/uadk/Makefile.am b/uadk/Makefile.am
--- a/uadk/Makefile.am	2025-12-23 13:54:34.000000000 +0800
+++ b/uadk/Makefile.am	2025-12-30 16:20:23.374816400 +0800
@@ -214,4 +214,3 @@
 pkgconfig_DATA = lib/libwd_crypto.pc lib/libwd_comp.pc lib/libwd.pc
 CLEANFILES += $(pkgconfig_DATA)
 
-SUBDIRS=. test v1/test uadk_tool sample
diff -urN a/uadk/cleanup.sh b/uadk/cleanup.sh
--- a/uadk/cleanup.sh	2025-12-23 13:54:34.000000000 +0800
+++ b/uadk/cleanup.sh	2025-12-30 16:19:34.378681100 +0800
@@ -10,8 +10,7 @@
        ar-lib m4 Makefile.in missing src/Makefile src/Makefile.in	     \
        test/Makefile test/Makefile.in test/hisi_hpre_test/Makefile.in	     \
        test/hisi_hpre_test/Makefile test/hisi_sec_test/Makefile              \
-       test/hisi_sec_test/Makefile.in  \
-       uadk_tool/Makefile uadk_tool/Makefile.in         \
+       test/hisi_sec_test/Makefile.in            \
        v1/Makefile.in v1/Makefile.in v1/test/Makefile v1/test/Makefile.in    \
        v1/test/test_mm/Makefile v1/test/test_mm/Makefile.in		     \
        v1/test/bmm_test/Makefile v1/test/bmm_test/Makefile.in"
diff -urN a/uadk/configure.ac b/uadk/configure.ac
--- a/uadk/configure.ac	2025-12-23 13:54:34.000000000 +0800
+++ b/uadk/configure.ac	2025-12-30 16:20:11.446749500 +0800
@@ -103,7 +103,6 @@
 		 lib/libwd_crypto.pc lib/libwd_comp.pc lib/libwd.pc
 		 test/Makefile
 		 test/hisi_hpre_test/Makefile
-		 uadk_tool/Makefile
 		 sample/Makefile
 		 v1/test/Makefile
 		 v1/test/bmm_test/Makefile
@@ -111,6 +110,5 @@
 		 v1/test/hisi_hpre_test/Makefile
 		 v1/test/hisi_sec_test/Makefile
 		 v1/test/hisi_sec_test_sgl/Makefile
-		 v1/test/hisi_zip_test/Makefile
 		 v1/test/hisi_zip_test_sgl/Makefile])
 AC_OUTPUT
diff -urN a/uadk/v1/test/Makefile.am b/uadk/v1/test/Makefile.am
--- a/uadk/v1/test/Makefile.am	2025-12-23 13:54:34.000000000 +0800
+++ b/uadk/v1/test/Makefile.am	2025-12-30 16:16:43.534866500 +0800
@@ -2,9 +2,6 @@
 
 SUBDIRS=. test_mm bmm_test
 
-if HAVE_ZLIB
-SUBDIRS+=hisi_zip_test
-endif
 
 SUBDIRS+=hisi_zip_test_sgl
 
