diff --git a/uadk/Makefile.am b/uadk/Makefile.am
index 9b36ae9..0df1dfd 100644
--- a/uadk/Makefile.am
+++ b/uadk/Makefile.am
@@ -41,6 +41,7 @@ pkginclude_HEADERS = include/wd.h include/wd_cipher.h include/wd_aead.h \
 nobase_pkginclude_HEADERS = v1/wd.h v1/wd_cipher.h v1/wd_aead.h v1/uacce.h v1/wd_dh.h \
 			 v1/wd_digest.h v1/wd_rsa.h v1/wd_bmm.h
 
+nobase_pkginclude_HEADERS += v1/wd_sgl.h
 lib_LTLIBRARIES=libwd.la libwd_comp.la libwd_crypto.la libwd_dae.la
 
 uadk_driversdir=$(libdir)/uadk
diff --git a/uadk/v1/drv/hisi_qm_udrv.c b/uadk/v1/drv/hisi_qm_udrv.c
index 7b0183b..224aadb 100644
--- a/uadk/v1/drv/hisi_qm_udrv.c
+++ b/uadk/v1/drv/hisi_qm_udrv.c
@@ -53,11 +53,11 @@ static int qm_hw_sgl_info(struct hw_sgl_info *sgl_info)
 
 /* 'num' starts from 1 */
 static int qm_hw_sgl_sge_init(struct wd_sgl *sgl, struct hisi_sgl *hisi_sgl,
-			      struct wd_mm_br *br, int num, __u32 buf_sz)
+			      struct wd_mm_br *br, int num)
 {
 	void *buf;
-
-	buf = wd_get_sge_buf(sgl, num);
+    __u32 buf_sz = sgl->sge[num - 1].data_len;
+	buf = sgl->sge[num - 1].buf;
 	if (!buf)
 		return -WD_EINVAL;
 
@@ -69,16 +69,17 @@ static int qm_hw_sgl_sge_init(struct wd_sgl *sgl, struct hisi_sgl *hisi_sgl,
 	}
 
 	hisi_sgl->sge_entries[num - 1].len = buf_sz;
-	drv_set_sgl_sge_pri(sgl, num - 1, &hisi_sgl->sge_entries[num - 1]);
+	sgl->sge[num - 1].priv = &hisi_sgl->sge_entries[num - 1];
 
 	return WD_SUCCESS;
 }
 
 /* 'num' starts from 1 */
 static void qm_hw_sgl_sge_uninit(struct wd_sgl *sgl, struct hisi_sgl *hisi_sgl,
-			  int num, struct wd_mm_br *br, __u32 buf_sz)
+			  int num, struct wd_mm_br *br)
 {
 	void *buf;
+    __u32 buf_sz = sgl->sge[num - 1].data_len;
 
 	buf = wd_get_sge_buf(sgl, num);
 	if (!buf)
@@ -90,24 +91,17 @@ static void qm_hw_sgl_sge_uninit(struct wd_sgl *sgl, struct hisi_sgl *hisi_sgl,
 
 static int qm_hw_sgl_init(void *pool, struct wd_sgl *sgl)
 {
-	int buf_num = wd_get_sgl_buf_num(sgl);
-	int sge_num = wd_get_sgl_sge_num(sgl);
+	int buf_num = sgl->buf_num;
+	int sge_num = sgl->sge_num;
 	struct hisi_sgl *hisi_sgl;
 	struct wd_mm_br *br;
 	int i, j, ret;
-	__u32 buf_sz;
 
 	if (!pool || buf_num < 0 || sge_num < 0) {
 		WD_ERR("hw_sgl_init init param err!\n");
 		return -WD_EINVAL;
 	}
 
-	ret = wd_get_sgl_bufsize(sgl, &buf_sz);
-	if (ret) {
-		WD_ERR("failed to get sgl bufsize!\n");
-		return ret;
-	}
-
 	buf_num = MIN(buf_num, HISI_SGL_SGE_NUM_MAX);
 	sge_num = MIN(sge_num, HISI_SGL_SGE_NUM_MAX);
 	br = drv_get_br(pool);
@@ -123,7 +117,7 @@ static int qm_hw_sgl_init(void *pool, struct wd_sgl *sgl)
 	hisi_sgl->next_dma = 0;
 
 	for (i = 0; i < buf_num; i++) {
-		ret = qm_hw_sgl_sge_init(sgl, hisi_sgl, br, i + 1, buf_sz);
+		ret = qm_hw_sgl_sge_init(sgl, hisi_sgl, br, i + 1);
 		if (ret) {
 			WD_ERR("failed to map buf, ret = %d.\n", ret);
 			goto sgl_sge_init_err;
@@ -136,12 +130,12 @@ static int qm_hw_sgl_init(void *pool, struct wd_sgl *sgl)
 		drv_set_sgl_sge_pri(sgl, i, &hisi_sgl->sge_entries[i]);
 	}
 
-	drv_set_sgl_pri(sgl, hisi_sgl);
+	sgl->priv = hisi_sgl;
 	return WD_SUCCESS;
 
 sgl_sge_init_err:
 	for (j = i - 1; j >= 0; j--)
-		qm_hw_sgl_sge_uninit(sgl, hisi_sgl, j + 1, br, buf_sz);
+		qm_hw_sgl_sge_uninit(sgl, hisi_sgl, j + 1, br);
 
 	br->free(br->usr, hisi_sgl);
 
@@ -609,10 +603,15 @@ int qm_send(struct wd_queue *q, void **req, __u32 num)
 	int ret;
 	__u32 i;
 
-	wd_fair_lock(&info->sd_lock);
+    if (!q->capa.priv.is_single_thread) {
+        wd_fair_lock(&info->sd_lock);
+    }
+
 	if (unlikely((__u32)__atomic_load_n(&info->used, __ATOMIC_RELAXED) >
-		     info->sq_depth - num - 1)) {
-		wd_fair_unlock(&info->sd_lock);
+		         info->sq_depth - num - 1)) {
+        if (!q->capa.priv.is_single_thread)
+            wd_fair_unlock(&info->sd_lock);
+
 		WD_ERR("queue is full!\n");
 		return -WD_EBUSY;
 	}
@@ -621,7 +620,9 @@ int qm_send(struct wd_queue *q, void **req, __u32 num)
 		ret = info->sqe_fill[qinfo->atype](req[i], qinfo->priv,
 				info->sq_tail_index);
 		if (unlikely(ret != WD_SUCCESS)) {
-			wd_fair_unlock(&info->sd_lock);
+            if (!q->capa.priv.is_single_thread)
+			    wd_fair_unlock(&info->sd_lock);
+
 			WD_ERR("sqe fill error, ret %d!\n", ret);
 			return -WD_EINVAL;
 		}
@@ -633,7 +634,8 @@ int qm_send(struct wd_queue *q, void **req, __u32 num)
 	}
 
 	ret = qm_tx_update(info, num);
-	wd_fair_unlock(&info->sd_lock);
+    if (!q->capa.priv.is_single_thread)
+	    wd_fair_unlock(&info->sd_lock);
 
 	return ret;
 }
@@ -686,7 +688,7 @@ int qm_rx_update(struct qm_queue_info *info, __u32 num)
 		return ret;
 
 	/* make sure queue status check is complete. */
-	rmb();
+	// rmb();
 
 	/* set c_flag to enable interrupt when use poll */
 	info->db(info, DOORBELL_CMD_CQ, info->cq_head_index, info->is_poll);
@@ -709,7 +711,9 @@ int qm_recv(struct wd_queue *q, void **resp, __u32 num)
 	if (unlikely(ret))
 		return ret;
 
-	wd_fair_lock(&info->rc_lock);
+    if (!q->capa.priv.is_single_thread)
+	    wd_fair_lock(&info->rc_lock);
+
 	for (i = 0; i < num; i++) {
 		cqe = info->cq_base + info->cq_head_index * sizeof(struct cqe);
 		if (info->cqc_phase != CQE_PHASE(cqe))
@@ -718,7 +722,9 @@ int qm_recv(struct wd_queue *q, void **resp, __u32 num)
 		mb(); /* make sure the data is all in memory before read */
 		sq_head = CQE_SQ_HEAD_INDEX(cqe);
 		if (unlikely(sq_head >= info->sq_depth)) {
-			wd_fair_unlock(&info->rc_lock);
+            if (!q->capa.priv.is_single_thread)
+			    wd_fair_unlock(&info->rc_lock);
+
 			WD_ERR("CQE_SQ_HEAD_INDEX(%u) error\n", sq_head);
 			return -WD_EIO;
 		}
@@ -730,7 +736,9 @@ int qm_recv(struct wd_queue *q, void **resp, __u32 num)
 		if (!ret) {
 			break;
 		} else if (ret < 0) {
-			wd_fair_unlock(&info->rc_lock);
+            if (!q->capa.priv.is_single_thread)
+			    wd_fair_unlock(&info->rc_lock);
+
 			WD_ERR("recv sqe error %u\n", sq_head);
 			return ret;
 		}
@@ -751,7 +759,8 @@ int qm_recv(struct wd_queue *q, void **resp, __u32 num)
 			ret = i;
 	}
 
-	wd_fair_unlock(&info->rc_lock);
+    if (!q->capa.priv.is_single_thread)
+	    wd_fair_unlock(&info->rc_lock);
 
 	return ret;
 }
diff --git a/uadk/v1/libwd.map b/uadk/v1/libwd.map
index d53201b..a43a884 100644
--- a/uadk/v1/libwd.map
+++ b/uadk/v1/libwd.map
@@ -139,6 +139,8 @@ global:
 	wcrypto_rng_poll;
 
 	wd_sglpool_create;
+    wd_build_sgl;
+    wd_destory_sgl;
 	wd_sglpool_destroy;
 	wd_alloc_sgl;
 	wd_free_sgl;
diff --git a/uadk/v1/wd.h b/uadk/v1/wd.h
index 4618a8c..51a2061 100644
--- a/uadk/v1/wd.h
+++ b/uadk/v1/wd.h
@@ -80,6 +80,7 @@ struct wcrypto_paras {
 	 */
 	__u8 direction;
 	__u8 is_poll;
+    __u8 is_single_thread;
 
 	 /* to be extended */
 };
diff --git a/uadk/v1/wd_comp.c b/uadk/v1/wd_comp.c
index 169f1b4..209fab0 100644
--- a/uadk/v1/wd_comp.c
+++ b/uadk/v1/wd_comp.c
@@ -253,16 +253,22 @@ int wcrypto_do_comp(void *ctx, struct wcrypto_comp_op_data *opdata, void *tag)
 	struct wcrypto_comp_ctx *cctx = ctx;
 	struct wcrypto_comp_msg *msg, *resp;
 	__u64 recv_count = 0;
-	int ret;
+	int ret = 0;
 
 	if (unlikely(!ctx || !opdata || !opdata->in || !opdata->out)) {
 		WD_ERR("invalid: comp input parameter err!\n");
 		return -EINVAL;
 	}
 
-	ret = wd_get_cookies(&cctx->pool, (void **)&cookie, 1);
-	if (ret)
-		return ret;
+   if (!opdata->cookie || !cctx->q->capa.priv.is_single_thread) {
+        ret = wd_get_cookies(&cctx->pool, (void **)&cookie, 1);
+        if (ret)
+            return ret;
+
+        opdata->cookie = cookie;
+    } else {
+        cookie = opdata->cookie;
+    }
 
 	msg = &cookie->msg;
 	if (tag) {
@@ -314,6 +320,7 @@ int wcrypto_do_comp(void *ctx, struct wcrypto_comp_op_data *opdata, void *tag)
 
 err_put_cookie:
 	wd_put_cookies(&cctx->pool, (void **)&cookie, 1);
+    opdata->cookie = NULL;
 	return ret;
 }
 
@@ -355,7 +362,9 @@ int wcrypto_comp_poll(struct wd_queue *q, unsigned int num)
 		tag = (void *)(uintptr_t)resp->udata;
 		ctx = tag->wcrypto_tag.ctx;
 		ctx->cb(resp, tag->wcrypto_tag.tag);
-		wd_put_cookies(&ctx->pool, (void **)&tag, 1);
+        if (!q->capa.priv.is_single_thread)
+            wd_put_cookies(&ctx->pool, (void **)&tag, 1);
+
 		resp = NULL;
 	} while (--tmp);
 
diff --git a/uadk/v1/wd_comp.h b/uadk/v1/wd_comp.h
index 4c84ea3..8c0d9e5 100644
--- a/uadk/v1/wd_comp.h
+++ b/uadk/v1/wd_comp.h
@@ -168,6 +168,7 @@ struct wcrypto_comp_op_data {
 	__u32 isize;
 	__u32 checksum;
 	void *priv;
+    void *cookie;
 };
 
 struct wcrypto_comp_msg {
diff --git a/uadk/v1/wd_sgl.c b/uadk/v1/wd_sgl.c
index cb3b8ee..fc1f329 100644
--- a/uadk/v1/wd_sgl.c
+++ b/uadk/v1/wd_sgl.c
@@ -44,30 +44,6 @@
 #define ALIGN_SIZE_MAX	0x800
 #define ALIGN_SIZE	64
 
-struct wd_sge {
-	/* 'priv' is used by driver, which may be a hardware sgl address */
-	void *priv;
-	void *buf;
-	__u32 data_len;
-	__u32 flag;
-	void *sgl;
-};
-
-struct wd_sgl {
-	/* 'priv' is hardware sgl address */
-	void *priv;
-	__u8 sge_num;
-	__u8 buf_num;
-	__u16 buf_sum;
-	__u32 sum_data_bytes;
-
-	struct wd_sglpool *pool;
-	struct wd_sgl *next;
-
-	/* user configuration, 60 sges max */
-	struct wd_sge sge[];
-};
-
 struct wd_sglpool {
 	struct wd_queue *q;
 	struct wd_lock sgl_lock;
@@ -167,6 +143,7 @@ static int sgl_chain_build(struct wd_queue *q, struct wd_sglpool *pool)
 				goto alloc_buf_err;
 			}
 			sgl_sge_init(sgl_blk[i], j, buf);
+            sgl_blk[i]->sge[j].data_len = sp->buf_size;
 		}
 
 		ret = drv_init_sgl(q, pool, sgl_blk[i]);
@@ -1029,3 +1006,76 @@ void wd_sgl_memset(struct wd_sgl *sgl, int ch)
 	for (i = 0; i < sgl->buf_num; i++)
 		memset(sgl->sge[i].buf, ch, sgl->pool->setup.buf_size);
 }
+
+struct mix_pool {
+    void *buf_pool;
+    void *usr_pool;
+    struct wd_sgl *sgl;
+    wd_map iova_map;
+};
+
+static void *wd_alloc_blk_in_build_sgl(void *pool, size_t size)
+{
+    struct mix_pool *mix_pool = pool;
+
+    if (mix_pool->sgl->priv)
+        return mix_pool->sgl->priv;
+
+    return wd_alloc_blk_sgl(mix_pool->buf_pool, size);
+}
+
+static void wd_free_blk_in_build_sgl(void *pool, void *blk)
+{
+    struct mix_pool *mix_pool = pool;
+
+    return wd_free_blk(mix_pool->buf_pool, blk);
+}
+
+static void *wd_iova_map_in_build_sgl(void *pool, void *va, size_t size)
+{
+    struct mix_pool *mix_pool = pool;
+    return mix_pool->iova_map(mix_pool->usr_pool, va, size);
+}
+
+int wd_build_sgl(struct wd_queue *q, void *pool, struct wd_sgl *sgl, const struct wd_buf_list *buf_list, wd_map usr_map)
+{
+    int ret = -WD_ENOMEM;
+    struct wd_sglpool *sgl_pool = pool;
+
+    sgl->buf_num = buf_list->buf_num;
+    sgl->sge_num = buf_list->buf_num;
+    sgl->pool = pool;
+    for (int i = 0; i < sgl->buf_num; i++) {
+        sgl->sge[i].flag = 0;
+        sgl->sge[i].buf = buf_list->buf[i].data;
+        sgl->sge[i].data_len = buf_list->buf[i].buf_len;
+        sgl->sge[i].sgl = sgl;
+    }
+
+    sgl->sge[sgl->buf_num - 1].flag |= FLAG_SGE_END;
+
+    struct mix_pool mix_pool;
+    mix_pool.buf_pool = sgl_pool->buf_br.usr;
+    mix_pool.usr_pool = buf_list->usr_data;
+    mix_pool.iova_map = usr_map;
+    mix_pool.sgl = sgl;
+    sgl_pool->buf_br.iova_map = wd_iova_map_in_build_sgl;
+    sgl_pool->buf_br.alloc = wd_alloc_blk_in_build_sgl;
+    sgl_pool->buf_br.free = wd_free_blk_in_build_sgl;
+    sgl_pool->buf_br.usr = &mix_pool;
+    ret = drv_init_sgl(q, pool, sgl);
+    sgl_pool->buf_br.alloc = wd_alloc_blk_sgl;
+    sgl_pool->buf_br.free = wd_free_blk;
+    sgl_pool->buf_br.iova_map = wd_blk_iova_map_sgl;
+    sgl_pool->buf_br.usr = mix_pool.buf_pool;
+    if (ret) {
+        WD_ERR("initialize hardware sgl failed, ret = %d\n", ret);
+    }
+
+	return ret;
+}
+
+int wd_destory_sgl(struct wd_queue *q, void *pool, struct wd_sgl *sgl)
+{
+    return drv_uninit_sgl(q, pool, sgl);
+}
diff --git a/uadk/v1/wd_sgl.h b/uadk/v1/wd_sgl.h
index e2e82f2..e254daa 100644
--- a/uadk/v1/wd_sgl.h
+++ b/uadk/v1/wd_sgl.h
@@ -43,7 +43,45 @@ struct wd_sglpool_setup {
 	struct wd_mm_br br;
 };
 
+struct wd_buf {
+    size_t buf_len;
+    void *data;
+};
+
+struct wd_buf_list {
+    unsigned int buf_num;
+    unsigned int rsvd;
+    struct wd_buf *buf;
+    void *usr_data;
+};
+
+struct wd_sge {
+	/* 'priv' is used by driver, which may be a hardware sgl address */
+	void *priv;
+	void *buf;
+	__u32 data_len;
+	__u32 flag;
+	void *sgl;
+};
+struct wd_sglpool;
+struct wd_sgl {
+	/* 'priv' is hardware sgl address */
+	void *priv;
+	__u8 sge_num;
+	__u8 buf_num;
+	__u16 buf_sum;
+	__u32 sum_data_bytes;
+
+	struct wd_sglpool *pool;
+	struct wd_sgl *next;
+
+	/* user configuration, 60 sges max */
+	struct wd_sge sge[];
+};
+
 void *wd_sglpool_create(struct wd_queue *q, struct wd_sglpool_setup *setup);
+int wd_build_sgl(struct wd_queue *q, void *pool, struct wd_sgl *sgl, const struct wd_buf_list *buf_list, wd_map usr_map);
+int wd_destory_sgl(struct wd_queue *q, void *pool, struct wd_sgl *sgl);
 void wd_sglpool_destroy(void *pool);
 struct wd_sgl *wd_alloc_sgl(void *pool, __u32 size);
 void wd_free_sgl(void *pool, struct wd_sgl *sgl);
